services/oauth/v2/oauth.go

```diff
@@ -0,0 +1,384 @@
+package v2
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/tidwall/gjson"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	rudderSync "github.com/rudderlabs/rudder-go-kit/sync"
+	obskit "github.com/rudderlabs/rudder-observability-kit/go/labels"
+	backendconfig "github.com/rudderlabs/rudder-server/backend-config"
+	router_utils "github.com/rudderlabs/rudder-server/router/utils"
+)
+
+func WithCache(cache Cache) func(*OAuthHandler) {
+	return func(h *OAuthHandler) {
+		h.Cache = cache
+	}
+}
+
+func WithLocker(lock *rudderSync.PartitionRWLocker) func(*OAuthHandler) {
+	return func(h *OAuthHandler) {
+		h.CacheMutex = lock
+	}
+}
+
+func WithExpirationTimeDiff(expirationTimeDiff time.Duration) func(*OAuthHandler) {
+	return func(h *OAuthHandler) {
+		h.ExpirationTimeDiff = expirationTimeDiff
+	}
+}
+
+func WithLogger(parentLogger logger.Logger) func(*OAuthHandler) {
+	return func(h *OAuthHandler) {
+		h.Logger = parentLogger
+	}
+}
+
+// NewOAuthHandler returns a new instance of OAuthHandler
+func NewOAuthHandler(provider TokenProvider, options ...func(*OAuthHandler)) *OAuthHandler {
+	cpTimeoutDuration := config.GetDuration("HttpClient.oauth.timeout", 30, time.Second)
+	h := &OAuthHandler{
+		TokenProvider: provider,
+		// This timeout is kind of modifiable & it seemed like 10 mins for this is too much!
+		RudderFlowType:            RudderFlow_Delivery,
+		AuthStatusUpdateActiveMap: make(map[string]bool),
+		ConfigBEURL:               backendconfig.GetConfigBackendURL(),
+		LoggerName:                "OAuthHandler",
+	}
+	for _, opt := range options {
+		opt(h)
+	}
+	if h.Logger == nil {
+		h.Logger = logger.NewLogger().Child("OAuthHandler")
+	}
+	h.CpConn = NewControlPlaneConnector(
+		WithCpClientTimeout(cpTimeoutDuration),
+		WithParentLogger(h.Logger),
+		WithLoggerName("OAuthHandler"),
+	)
```

Comment:
We can simply create a child logger instead of setting parent logger and logger name.
```suggestion
	h.CpConn = NewControlPlaneConnector(
		WithCpClientTimeout(conf.GetDuration("HttpClient.oauth.timeout", 30, time.Second)),
		WithCpLogger(h.Logger.Child("OAuthHandler")),
	)
```

Reply:
we don't need these. 
```GO
func WithParentLogger(parentLogger logger.Logger) func(*controlPlaneConnector) {
	return func(cpConn *controlPlaneConnector) {
		cpConn.logger = parentLogger
	}
}

func WithLoggerName(loggerName string) func(*controlPlaneConnector) {
	return func(h *controlPlaneConnector) {
		h.loggerName = loggerName
	}
}
```

Instead we can simply have one:
```GO
func WithCpLogger(logger logger.Logger) func(*controlPlaneConnector) {
	return func(cpConn *controlPlaneConnector) {
		cpConn.logger = logger
	}
}
```